File: smart-autonomous-vehicle/backend/control/vehicle_controller.py
"""
Longitudinal speed and emergency brake controller.
"""

from typing import Tuple, Optional
import numpy as np

try:
    from ..config import Config, get_config
except (ImportError, ValueError):
    try:
        from backend.config import Config, get_config
    except (ImportError, ValueError):
        from config import Config, get_config


def vehicle_controller(
    target_velocity: float,
    current_velocity: float,
    risk_level: str,
    cfg: Optional[Config] = None,
) -> Tuple[float, float, float]:
    """
    Computes throttle, brake, and net acceleration commands.
    """
    if cfg is None:
        cfg = get_config()

    max_accel = cfg.vehicle.max_accel
    max_decel = cfg.vehicle.max_decel
    kp = cfg.control.speed_kp

    if str(risk_level).upper() == "CRITICAL":
        return 0.0, 1.0, -max_decel

    eff_target_vel = target_velocity
    if str(risk_level).upper() == "WARNING":
        eff_target_vel = min(target_velocity, 3.5)

    vel_error = eff_target_vel - current_velocity
    desired_accel = kp * vel_error

    if desired_accel >= 0:
        throttle_cmd = min(1.0, desired_accel / max_accel)
        brake_cmd = 0.0
        accel_cmd = min(max_accel, desired_accel)
    else:
        throttle_cmd = 0.0
        brake_cmd = min(1.0, abs(desired_accel) / max_decel)
        accel_cmd = max(-max_decel, desired_accel)

    return float(throttle_cmd), float(brake_cmd), float(accel_cmd)

File: control/vehicleController.m
function [throttleCmd, brakeCmd, accelCmd] = vehicleController(targetVelocity, currentVelocity, riskLevel, cfg)
% VEHICLECONTROLLER Longitudinal velocity & emergency braking controller for autonomous vehicle.
%
% Inputs:
%   targetVelocity  - Target forward velocity from path planner (m/s)
%   currentVelocity - Current forward speed (m/s)
%   riskLevel       - "SAFE" | "WARNING" | "CRITICAL"
%   cfg             - System configuration struct
%
% Outputs:
%   throttleCmd - Throttle command in [0, 1]
%   brakeCmd    - Brake command in [0, 1]
%   accelCmd    - Net acceleration command (m/s^2)

    if nargin < 4 || isempty(cfg)
        cfg = config();
    end

    maxAccel = cfg.vehicle.maxAccel;
    maxDecel = cfg.vehicle.maxDecel;
    kp = cfg.control.speedKp;

    % Hard override for CRITICAL collision risk
    if string(riskLevel) == "CRITICAL"
        throttleCmd = 0.0;
        brakeCmd    = 1.0;
        accelCmd    = -maxDecel;
        return;
    end

    % Adjust target velocity for WARNING risk
    effTargetVel = targetVelocity;
    if string(riskLevel) == "WARNING"
        effTargetVel = min(targetVelocity, 3.5); % Slow down to cautious crawling speed
    end

    % Velocity error
    velError = effTargetVel - currentVelocity;

    % Proportional acceleration control
    desiredAccel = kp * velError;

    if desiredAccel >= 0
        % Accelerating
        throttleCmd = min(1.0, desiredAccel / maxAccel);
        brakeCmd = 0.0;
        accelCmd = min(maxAccel, desiredAccel);
    else
        % Braking / Coasting
        throttleCmd = 0.0;
        brakeCmd = min(1.0, abs(desiredAccel) / maxDecel);
        accelCmd = max(-maxDecel, desiredAccel);
    end
end
